File: backend/src/models/ScrapingTask.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const ScrapingTask = sequelize.define('ScrapingTask', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
    allowNull: false,
  },
  task_id: {
    type: DataTypes.STRING(20),
    allowNull: false,
    unique: true,
  },
  user_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  status: {
    type: DataTypes.STRING(20),
    defaultValue: 'PENDING',
  },
  location: {
    type: DataTypes.STRING(255),
    allowNull: false,
  },
  keyword: {
    type: DataTypes.STRING(255),
    allowNull: false,
  },
  search_radius: {
    type: DataTypes.INTEGER,
    defaultValue: 10,
  },
  max_results: {
    type: DataTypes.INTEGER,
    defaultValue: 100,
  },
  max_pages_per_site: {
    type: DataTypes.INTEGER,
    defaultValue: 20,
  },
  required_fields: {
    type: DataTypes.TEXT,
    defaultValue: '[]',
  },
  results_discovered: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  websites_found: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  websites_crawled: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  phones_found: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  emails_found: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  addresses_found: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  duplicates_removed: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  error_message: {
    type: DataTypes.TEXT,
    allowNull: true,
  },
  created_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW,
  },
  updated_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW,
  },
  completed_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
}, {
  tableName: 'scraping_tasks',
  timestamps: false,
});

module.exports = ScrapingTask;

File: backend/app/models/scraping_task.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON, func
from sqlalchemy.orm import relationship
from app.database.database import Base


class ScrapingTask(Base):
    __tablename__ = "scraping_tasks"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(String(20), unique=True, nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    status = Column(String(20), default="PENDING")  # PENDING/RUNNING/COMPLETED/FAILED/CANCELLED

    location = Column(String(255), nullable=False)
    keyword = Column(String(255), nullable=False)
    search_radius = Column(Integer, default=10)
    max_results = Column(Integer, default=100)
    max_pages_per_site = Column(Integer, default=20)
    required_fields = Column(Text, default="[]")  # JSON string

    # Progress counters
    results_discovered = Column(Integer, default=0)
    websites_found = Column(Integer, default=0)
    websites_crawled = Column(Integer, default=0)
    phones_found = Column(Integer, default=0)
    emails_found = Column(Integer, default=0)
    addresses_found = Column(Integer, default=0)
    duplicates_removed = Column(Integer, default=0)

    error_message = Column(Text, nullable=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    completed_at = Column(DateTime, nullable=True)

    user = relationship("User", back_populates="tasks")
    organizations = relationship("Organization", back_populates="task", cascade="all, delete-orphan")
    logs = relationship("ScrapingLog", back_populates="task", cascade="all, delete-orphan")
    exports = relationship("Export", back_populates="task")
